### PR TITLE
cmake: New INSTALL.md for build and testing instructions.

### PR DESCRIPTION
**Pull request by aniemetz**

### CODE REVIEW

File: INSTALL.md [hunk 1]
```diff
@@ -0,0 +1,326 @@
...
+bit-blasting. This dependency may improve performance.
+It can be installed using the `contrib/get-cryptominisat` script.  
+Configure CVC4 with `configure.sh --cryptominisat` to build with this
+dependency.
+
+### LFSC (The LFSC Proof Checker)
+
+[LFSC](https://github.com/CVC4/LFSC) is required to check proofs internally
+with --check-proofs. It can be installed using the `contrib/get-lfsc` script.  
+Configure CVC4 with `configure.sh --proofs` to build with this dependency.
```
Comment (4tXJ7f) [inline INSTALL.md]: Isn't the option `--lfsc`?

File: INSTALL.md [hunk 2]
```diff
@@ -0,0 +1,326 @@
...
+
+### CaDiCaL (Optional SAT solver)
+
+[CaDiCaL](https://github.com/arminbiere/cadical)
+is a SAT solver that can be used for solving non-incremental bit-vector
+problems with eager bit-blasting. This dependency may improve performance.
+It can be installed using the `contrib/get-cadical script`.  
+Configure CVC4 with `configure.sh --cadical` to build with this dependency.
+
+### CryptoMiniSat (optional SAT solver)
```
Comment (4tXJ7f) [inline INSTALL.md]: Nit: Capital "O" to match "CaDiCaL (Optional SAT solver)"

File: INSTALL.md [hunk 3]
```diff
@@ -0,0 +1,326 @@
...
+### Warning: GCC 4.5.1
+
+GCC version 4.5.1 seems to have a bug in the optimizer that may result in
+incorrect behavior (and wrong results) in many builds. This is a known problem
+for MiniSat, and since MiniSat is at the core of CVC4, a problem for CVC4.
+We recommend using a GCC version > 4.5.1.
+
+## Optional Dependencies
+
+### SymFPU (Support for the Theory of Floating Points)
```
Comment (4tXJ7f) [inline INSTALL.md]: Nit: "Floating Points" -> "Floating Point Numbers"

File: examples/README [hunk 4]
```diff
@@ -1,33 +1,31 @@
+Examples
+--------
```
Comment (4tXJ7f) [inline examples/README]: Suggestion, use `# Examples` for consistency.

File: INSTALL.md [hunk 5]
```diff
@@ -0,0 +1,326 @@
+CVC4 prerelease version 1.7.
+============================
```
Comment (4tXJ7f) [inline INSTALL.md]: Suggestion: use `#` for consistency instead of `====`.

File: INSTALL.md [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+CVC4 can be built on Linux and macOS.  For Windows, CVC4 can be cross-compiled
+using Mingw-w64.  We recommend a 64-bit operating system.
+
+On macOS, we recommend using Homebrew (https://brew.sh/) to install the
+dependencies.  We also have a Homebrew Tap available at
+https://github.com/CVC4/homebrew-cvc4 .
+
+## Build dependencies
+
+The following tools and libraries are required to build and run CVC4.
```
Comment (HanielB) [inline INSTALL.md]: trailing whitespaces after period

File: INSTALL.md [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+
+## Optional Dependencies
+
+### SymFPU (Support for the Theory of Floating Points)
+
+[SymFPU](https://github.com/martin-cs/symfpu/tree/CVC4)
+is an implementation of SMT-LIB/IEEE-754 floating-point operations in terms
+of bit-vector operations.
+It is required for supporting the theory of floating-point numbers and
+can be installed using the `contrib/get-symfpu` script.
```
Comment (HanielB) [inline INSTALL.md]: ditto

File: INSTALL.md [hunk 8]
```diff
@@ -0,0 +1,326 @@
...
+It is required for supporting the theory of floating-point numbers and
+can be installed using the `contrib/get-symfpu` script.  
+Configure CVC4 with `configure.sh --symfpu` to build with this dependency.
+
+### CaDiCaL (Optional SAT solver)
+
+[CaDiCaL](https://github.com/arminbiere/cadical)
+is a SAT solver that can be used for solving non-incremental bit-vector
+problems with eager bit-blasting. This dependency may improve performance.
+It can be installed using the `contrib/get-cadical script`.
```
Comment (HanielB) [inline INSTALL.md]: ditto

File: INSTALL.md [hunk 9]
```diff
@@ -0,0 +1,326 @@
...
+problems with eager bit-blasting. This dependency may improve performance.
+It can be installed using the `contrib/get-cadical script`.  
+Configure CVC4 with `configure.sh --cadical` to build with this dependency.
+
+### CryptoMiniSat (optional SAT solver)
+
+[CryptoMinisat](https://github.com/msoos/cryptominisat)
+is a SAT solver that can be used for solving bit-vector problems with eager
+bit-blasting. This dependency may improve performance.
+It can be installed using the `contrib/get-cryptominisat` script.
```
Comment (HanielB) [inline INSTALL.md]: ditto

File: INSTALL.md [hunk 10]
```diff
@@ -0,0 +1,326 @@
...
+is a SAT solver that can be used for solving bit-vector problems with eager
+bit-blasting. This dependency may improve performance.
+It can be installed using the `contrib/get-cryptominisat` script.  
+Configure CVC4 with `configure.sh --cryptominisat` to build with this
+dependency.
+
+### LFSC (The LFSC Proof Checker)
+
+[LFSC](https://github.com/CVC4/LFSC) is required to check proofs internally
+with --check-proofs. It can be installed using the `contrib/get-lfsc` script.
```
Comment (HanielB) [inline INSTALL.md]: ditto

File: INSTALL.md [hunk 11]
```diff
@@ -0,0 +1,326 @@
...
+### SWIG >= 3.0.x (Simplified Wrapper and Interface Generator)
+
+SWIG 3.0.x (and a JDK) is necessary to build the Java API.
+See [Language Bindings](language-bindings) below for build instructions.
+
+### CLN >= v1.3 (Class Library for Numbers)
+
+[CLN](http://www.ginac.de/CLN)
+is an alternative multiprecision arithmetic package that may offer better
+performance and memory footprint than GMP.
```
Comment (HanielB) [inline INSTALL.md]: ditto

File: INSTALL.md [hunk 12]
```diff
@@ -0,0 +1,326 @@
...
+### glpk-cut-log (A fork of the GNU Linear Programming Kit)
+
+[glpk-cut-log](https://github.com/timothy-king/glpk-cut-log/) is a fork of
+[GLPK](http://www.gnu.org/software/glpk/) (the GNU Linear Programming Kit).
+This can be used to speed up certain classes of problems for the arithmetic
+implementation in CVC4. (This is not recommended for most users.)
+
+glpk-cut-log can be installed using the `contrib/get-glpk-cut-log` script.
+Note that the only installation option is manual installation via this script.
+CVC4 is no longer compatible with the main GLPK library.
```
Comment (HanielB) [inline INSTALL.md]: ditto

File: INSTALL.md [hunk 13]
```diff
@@ -0,0 +1,326 @@
...
+
+### ABC library (Improved Bit-Vector Support)
+
+[ABC](http://www.eecs.berkeley.edu/~alanmi/abc/) (A System for Sequential
+Synthesis and Verification) is a library for synthesis and verification of
+logic circuits. This dependency may improve performance of the eager
+bit-vector solver. When enabled, the bit-blasted formula is encoded into
+and-inverter-graphs (AIG) and ABC is used to simplify these AIGs.
+
+ABC can be installed using the `contrib/get-abc` script.
```
Comment (HanielB) [inline INSTALL.md]: ditto

File: INSTALL.md [hunk 14]
```diff
@@ -0,0 +1,326 @@
...
+`<build_dir>/src`.
+
+## Language bindings
+
+CVC4 provides a complete and flexible C++ API (see `examples/api` for examples).
+It further provides Java (see `examples/SimpleVC.java` and `examples/api/java`)
+and Python (see `examples/SimpleVC.py`) API bindings.
+
+Configure CVC4 with `configure.sh --language-bindings=[java,python,all]`
+to build with language bindings.
```
Comment (HanielB) [inline INSTALL.md]: ditto

File: INSTALL.md [hunk 15]
```diff
@@ -0,0 +1,326 @@
...
+  (label: **system**)
+- **unit tests** in directory `test/unit`
+  (label: **unit**)
+
+### Testing Examples
+
+For building instructions, see [Building the Examples](building-the-examples).
+
+We use prefix `example/` + `<subdir>/` + `<example>` (for `<example>` in
+`example/<subdir>/`) as test target name.
```
Comment (HanielB) [inline INSTALL.md]: ditto

File: INSTALL.md [hunk 16]
```diff
@@ -0,0 +1,326 @@
...
+The system tests are not built by default.
+
+    make systemtests                      # build and run all system tests
+    make <system_test>                    # build test/system/<system_test>.<ext>
+    ctest system/<system_test>            # run test/system/<system_test>.<ext>
+
+All system test binaries are built into `<build_dir>/bin/test/system`.
+
+We use prefix `system/` + `<system_test>` (for `<system_test>` in `test/system`)
+as test target name.
```
Comment (HanielB) [inline INSTALL.md]: ditto

File: INSTALL.md [hunk 17]
```diff
@@ -0,0 +1,326 @@
...
+The unit tests are not built by default.
+
+    make units                            # build and run all unit tests
+    make <unit_test>                      # build test/unit/<subdir>/<unit_test>.<ext>
+    ctest unit/<subdir>/<unit_test>       # run test/unit/<subdir>/<unit_test>.<ext>
+
+All unit test binaries are built into `<build_dir>/bin/test/unit`.
+
+We use prefix `unit/` + `<subdir>/` + `<unit_test>` (for `<unit_test>` in
+`test/unit/<subdir>`) as test target name.
```
Comment (HanielB) [inline INSTALL.md]: ditto

File: INSTALL.md [hunk 18]
```diff
@@ -0,0 +1,326 @@
...
+                                          # > runs unit/base/map_util_black
+    ctest -R base/map_util_black$         # run all tests that match '*base/map_util_black'
+                                          # > runs unit/base/map_util_black
+    ctest -R unit/base/map_util_black$    # run all tests that match '*unit/base/map_util_black'
+                                          # > runs unit/base/map_util_black
+
+### Testing Regression Tests
+
+We use prefix `regressN/` + `<subdir>/` + `<regress_test>` (for `<regress_test>`
+in level `N` in `test/regress/regressN/<subdir>`) as test target name.
```
Comment (HanielB) [inline INSTALL.md]: ditto

File: INSTALL.md [hunk 19]
```diff
@@ -0,0 +1,326 @@
...
+    ctest -R regress0                     # run all regression tests in level 0
+    ctest -R regress[0-1]                 # run all regression tests in level 0 and 1
+    ctest -R regress0/bug288b             # run all tests that match '*bug288b*'
+                                          # > runs regress0/bug288b
+### Custom Targets
+
+All custom test targets build and run a preconfigured set of tests and or
+examples.
+
+- `make check [-jN] [ARGS=-jN]`
```
Comment (HanielB) [inline INSTALL.md]: ditto (and after the other `make` cases below as well)

File: INSTALL.md [hunk 20]
```diff
@@ -0,0 +1,326 @@
...
+- [GNU C and C++ (gcc and g++)](https://gcc.gnu.org)
+  or [Clang](https://clang.llvm.org) (reasonably recent versions)
+- [CMake >= 3.1](https://cmake.org)
+- [GNU Bash](https://www.gnu.org/software/bash/)
+- [Python >= 2.7](https://www.python.org)
+- [GMP v4.2 (GNU Multi-Precision arithmetic library)](https://gmplib.org)
+- [libantlr3c v3.2 or v3.4 (ANTLR parser generator C support library)](http://www.antlr3.org/)
+
+Some features, such as the theory of floating-point numbers, require
+[optional dependencies](optional-dependencies) (see below).
```
[the remark this answers is not in the document]
Comment (aniemetz) [reply INSTALL.md]: We do since this is also included in the source release (and in text files, links do not work).

File: examples/README (hunk 4 above)
Comment (aniemetz) [reply examples/README]: This is on purpose and not the same as #.

File: INSTALL.md (hunk 5 above)
Comment (aniemetz) [reply INSTALL.md]: See above, this is on purpose and not the same as #.

File: INSTALL.md (hunk 6 above)
Comment (aniemetz) [reply INSTALL.md]: Trailing whitespaces are necessary for line breaks (markdown).

File: INSTALL.md (hunk 7 above)
Comment (aniemetz) [reply INSTALL.md]: Ditto

File: INSTALL.md (hunk 6 above)
Comment (HanielB) [reply INSTALL.md]: oh my god

File: INSTALL.md [hunk 21]
```diff
@@ -0,0 +1,326 @@
...
+- **system tests** in directory `test/system`
+  (label: **system**)
+- **unit tests** in directory `test/unit`
+  (label: **unit**)
+
+### Testing Examples
+
+For building instructions, see [Building the Examples](building-the-examples).
+
+We use prefix `example/` + `<subdir>/` + `<example>` (for `<example>` in
```
[the remark this answers is not in the document]
Comment (aniemetz) [reply INSTALL.md]: No, the prefix is `example/`.

File: INSTALL.md [hunk 22]
```diff
@@ -0,0 +1,326 @@
...
+`example/<subdir>/`) as test target name.  
+
+    make bitvectors                       # build example/api/bitvectors.cpp
```
[the remark this answers is not in the document]
Comment (aniemetz) [reply INSTALL.md]: As above.

**Pull request merged**